File: sdk/java/src/com/jicek/sdk/fingerprint/FingerprintCollector.java
package com.jicek.sdk.fingerprint;

import com.jicek.sdk.JicekException;
import com.jicek.sdk.crypto.CryptoUtil;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * 设备指纹采集器（5 维 SHA-256 融合）
 * 作者: 极策k  日期: 2026-07-21
 *
 * 5 维：CPU / 主板 / 硬盘 / 网卡 MAC / BIOS UUID
 * VM 场景：追加 VM UUID 或容器 ID 作为补充维度
 *
 * 平台支持：Windows / Linux / macOS
 * 采集失败时该维度返回空字符串（不禁整个流程）
 */
public class FingerprintCollector {

    private static final String OS = System.getProperty("os.name", "").toLowerCase();

    /**
     * 采集并计算最终指纹
     *
     * @return 指纹结果（fingerprint + encryptedDetail + isVm + vmExtra + 元信息）
     */
    public FingerprintResult collect(String rsaPublicKey) {
        Map<String, String> dimensions = new LinkedHashMap<>();
        dimensions.put("cpu", collectCpu());
        dimensions.put("mainboard", collectMainboard());
        dimensions.put("disk", collectDisk());
        dimensions.put("mac", collectMac());
        dimensions.put("bios", collectBios());

        // 5 维单独 SHA-256
        Map<String, String> hashedDims = new LinkedHashMap<>();
        for (Map.Entry<String, String> e : dimensions.entrySet()) {
            hashedDims.put(e.getKey(), CryptoUtil.sha256Hex(e.getValue()));
        }

        // VM/容器检测
        String vmExtra = detectVmExtra();
        boolean isVm = vmExtra != null && !vmExtra.isBlank();

        // 最终指纹 = SHA-256(5维哈希拼接 [+ vmExtra])
        StringBuilder fpInput = new StringBuilder();
        fpInput.append(hashedDims.get("cpu"));
        fpInput.append(hashedDims.get("mainboard"));
        fpInput.append(hashedDims.get("disk"));
        fpInput.append(hashedDims.get("mac"));
        fpInput.append(hashedDims.get("bios"));
        if (isVm) {
            fpInput.append(vmExtra);
        }
        String fingerprint = CryptoUtil.sha256Hex(fpInput.toString());

        // 5 维哈希 JSON → RSA 加密
        String detailJson = toJson(hashedDims);
        String encryptedDetail = CryptoUtil.rsaEncrypt(detailJson, rsaPublicKey);

        return new FingerprintResult(
                fingerprint,
                encryptedDetail,
                isVm ? 1 : 0,
                vmExtra,
                detectOsType(),
                detectOsVersion(),
                detectHostName(),
                detectClientVersion()
        );
    }

    /* ============ 5 维采集 ============ */

    private String collectCpu() {
        if (OS.contains("win")) {
            return execRead("wmic cpu get ProcessorId /value").replace("ProcessorId=", "").trim();
        } else if (OS.contains("mac")) {
            return execRead("sysctl -n machdep.cpu.brand_string");
        } else {
            return execRead("cat /proc/cpuinfo | grep -i serial | head -1")
                    .replaceAll("(?i).*serial\\s*:", "").trim();
        }
    }

    private String collectMainboard() {
        if (OS.contains("win")) {
            return execRead("wmic baseboard get SerialNumber /value")
                    .replace("SerialNumber=", "").trim();
        } else if (OS.contains("mac")) {
            return execRead("ioreg -l | grep IOPlatformSerialNumber | head -1")
                    .replaceAll(".*\"(.+)\".*", "$1");
        } else {
            return execRead("cat /sys/class/dmi/id/board_serial 2>/dev/null || dmidecode -s baseboard-serial-number 2>/dev/null");
        }
    }

    private String collectDisk() {
        if (OS.contains("win")) {
            return execRead("wmic diskdrive get SerialNumber /value")
                    .replace("SerialNumber=", "").trim();
        } else if (OS.contains("mac")) {
            return execRead("diskutil info / | grep 'Volume UUID' | awk '{print $3}'");
        } else {
            return execRead("cat /sys/block/sda/device/serial 2>/dev/null || hdparm -I /dev/sda 2>/dev/null | grep 'Serial No' | awk -F: '{print $2}'")
                    .trim();
        }
    }

    private String collectMac() {
        if (OS.contains("win")) {
            return execRead("getmac /fo csv /nh | head -1 | awk -F',' '{print $1}'")
                    .replace("\"", "").trim();
        } else if (OS.contains("mac")) {
            return execRead("ifconfig en0 | grep ether | awk '{print $2}'");
        } else {
            return execRead("cat /sys/class/net/$(ip route | grep default | awk '{print $5}' | head -1)/address")
                    .trim();
        }
    }

    private String collectBios() {
        if (OS.contains("win")) {
            return execRead("wmic bios get SerialNumber /value")
                    .replace("SerialNumber=", "").trim();
        } else if (OS.contains("mac")) {
            return execRead("ioreg -l | grep IOPlatformUUID | head -1")
                    .replaceAll(".*\"(.+)\".*", "$1");
        } else {
            return execRead("cat /sys/class/dmi/id/product_uuid 2>/dev/null || dmidecode -s system-uuid 2>/dev/null")
                    .trim();
        }
    }

    /* ============ VM/容器检测 ============ */

    private String detectVmExtra() {
        // Linux 容器检测
        if (OS.contains("linux")) {
            String cgroup = execRead("cat /proc/self/cgroup 2>/dev/null | grep docker | head -1");
            if (cgroup.contains("docker")) {
                return "container:" + cgroup.replaceAll(".*docker/([a-f0-9]+).*", "$1");
            }
            // VM UUID
            String vmUuid = execRead("dmidecode -s system-uuid 2>/dev/null").trim();
            if (!vmUuid.isEmpty() && !vmUuid.equals("Not Settable") && !vmUuid.equals("Not Specified")) {
                return "vm:" + vmUuid;
            }
        }
        // Windows VM 检测
        if (OS.contains("win")) {
            String model = execRead("wmic computersystem get Model /value")
                    .replace("Model=", "").trim();
            if (model.contains("Virtual") || model.contains("VMware") || model.contains("KVM")) {
                String uuid = execRead("wmic csproduct get UUID /value")
                        .replace("UUID=", "").trim();
                return "vm:" + uuid;
            }
        }
        return null;
    }

    /* ============ 元信息 ============ */

    private String detectOsType() {
        if (OS.contains("win")) return "windows";
        if (OS.contains("mac")) return "macos";
        if (OS.contains("linux")) return "linux";
        return OS;
    }

    private String detectOsVersion() {
        return System.getProperty("os.version", "");
    }

    private String detectHostName() {
        try {
            return java.net.InetAddress.getLocalHost().getHostName();
        } catch (Exception e) {
            return "unknown";
        }
    }

    private String detectClientVersion() {
        return "jicek-sdk-java-0.3.1";
    }

    /* ============ 工具方法 ============ */

    private String execRead(String cmd) {
        try {
            String[] shell;
            if (OS.contains("win")) {
                shell = new String[]{"cmd", "/c", cmd};
            } else {
                shell = new String[]{"/bin/sh", "-c", cmd};
            }
            Process p = new ProcessBuilder(shell)
                    .redirectErrorStream(true)
                    .start();
            String output;
            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(p.getInputStream(), StandardCharsets.UTF_8))) {
                output = reader.lines().collect(Collectors.joining("\n"));
            }
            p.waitFor();
            return output == null ? "" : output;
        } catch (Exception e) {
            return "";
        }
    }

    private String toJson(Map<String, String> map) {
        StringBuilder sb = new StringBuilder("{");
        boolean first = true;
        for (Map.Entry<String, String> e : map.entrySet()) {
            if (!first) sb.append(",");
            sb.append("\"").append(e.getKey()).append("\":\"").append(e.getValue()).append("\"");
            first = false;
        }
        sb.append("}");
        return sb.toString();
    }
}

File: sdk/csharp/JicekSdk/FingerprintCollector.cs
// 极策k C# SDK 设备指纹采集
// 作者: 极策k  日期: 2026-07-21
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;
using System.Text.Json;

namespace Jicek.Sdk
{
    public class FingerprintResult
    {
        public string Fingerprint { get; set; } = "";
        public string EncryptedDetail { get; set; } = "";
        public int IsVm { get; set; }
        public string VmExtra { get; set; } = "";
        public string OsType { get; set; } = "";
        public string OsVersion { get; set; } = "";
        public string DeviceName { get; set; } = "";
        public string ClientVersion { get; set; } = "";
    }

    public class FingerprintCollector
    {
        public FingerprintResult Collect(string rsaPublicKey)
        {
            var cpu = CollectCpu();
            var mainboard = CollectMainboard();
            var disk = CollectDisk();
            var mac = CollectMac();
            var bios = CollectBios();

            var hashed = new Dictionary<string, string>
            {
                ["cpu"] = CryptoUtil.Sha256Hex(cpu),
                ["mainboard"] = CryptoUtil.Sha256Hex(mainboard),
                ["disk"] = CryptoUtil.Sha256Hex(disk),
                ["mac"] = CryptoUtil.Sha256Hex(mac),
                ["bios"] = CryptoUtil.Sha256Hex(bios),
            };

            var vmExtra = DetectVmExtra();
            var isVm = string.IsNullOrEmpty(vmExtra) ? 0 : 1;

            var fpInput = hashed["cpu"] + hashed["mainboard"] + hashed["disk"] + hashed["mac"] + hashed["bios"];
            if (!string.IsNullOrEmpty(vmExtra)) fpInput += vmExtra;
            var fingerprint = CryptoUtil.Sha256Hex(fpInput);

            var detailJson = JsonSerializer.Serialize(hashed);
            var encryptedDetail = CryptoUtil.RsaEncryptOaep(detailJson, rsaPublicKey);

            return new FingerprintResult
            {
                Fingerprint = fingerprint,
                EncryptedDetail = encryptedDetail,
                IsVm = isVm,
                VmExtra = vmExtra,
                OsType = OsType(),
                OsVersion = Environment.OSVersion.VersionString,
                DeviceName = Environment.MachineName,
                ClientVersion = "jicek-sdk-csharp-0.3.1",
            };
        }

        private static string Exec(string cmd)
        {
            try
            {
                var psi = new ProcessStartInfo
                {
                    FileName = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "cmd" : "/bin/sh",
                    Arguments = (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "/c " : "-c ") + cmd,
                    RedirectStandardOutput = true,
                    UseShellExecute = false,
                    CreateNoWindow = true,
                };
                using var p = Process.Start(psi);
                if (p == null) return "";
                var output = p.StandardOutput.ReadToEnd();
                p.WaitForExit(3000);
                return output.Trim();
            }
            catch { return ""; }
        }

        private static string CollectCpu()
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                return Exec("wmic cpu get ProcessorId /value").Replace("ProcessorId=", "").Trim();
            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
                return Exec("sysctl -n machdep.cpu.brand_string");
            return Exec("cat /proc/cpuinfo | grep -i serial | head -1");
        }

        private static string CollectMainboard()
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                return Exec("wmic baseboard get SerialNumber /value").Replace("SerialNumber=", "").Trim();
            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
                return Exec("ioreg -l | grep IOPlatformSerialNumber | head -1");
            return Exec("cat /sys/class/dmi/id/board_serial 2>/dev/null || dmidecode -s baseboard-serial-number 2>/dev/null");
        }

        private static string CollectDisk()
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                return Exec("wmic diskdrive get SerialNumber /value").Replace("SerialNumber=", "").Trim();
            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
                return Exec("diskutil info / | grep 'Volume UUID' | awk '{print $3}'");
            return Exec("cat /sys/block/sda/device/serial 2>/dev/null || hdparm -I /dev/sda 2>/dev/null | grep 'Serial No' | awk -F: '{print $2}'").Trim();
        }

        private static string CollectMac()
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                var out_ = Exec("getmac /fo csv /nh");
                var idx = out_.IndexOf(',');
                return idx > 0 ? out_.Substring(0, idx).Trim('"') : "";
            }
            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
                return Exec("ifconfig en0 | grep ether | awk '{print $2}'");
            return Exec("cat /sys/class/net/$(ip route | grep default | awk '{print $5}' | head -1)/address").Trim();
        }

        private static string CollectBios()
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                return Exec("wmic bios get SerialNumber /value").Replace("SerialNumber=", "").Trim();
            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
                return Exec("ioreg -l | grep IOPlatformUUID | head -1");
            return Exec("cat /sys/class/dmi/id/product_uuid 2>/dev/null || dmidecode -s system-uuid 2>/dev/null").Trim();
        }

        private static string DetectVmExtra()
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            {
                var cgroup = Exec("cat /proc/self/cgroup 2>/dev/null | grep docker | head -1");
                if (cgroup.Contains("docker"))
                {
                    var idx = cgroup.IndexOf("docker/");
                    if (idx >= 0) return "container:" + cgroup.Substring(idx + 7).Split('\n')[0];
                }
                var vmUuid = Exec("dmidecode -s system-uuid 2>/dev/null").Trim();
                if (!string.IsNullOrEmpty(vmUuid) && vmUuid != "Not Settable" && vmUuid != "Not Specified")
                    return "vm:" + vmUuid;
            }
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                var model = Exec("wmic computersystem get Model /value").Replace("Model=", "").Trim();
                if (model.Contains("Virtual") || model.Contains("VMware") || model.Contains("KVM"))
                {
                    var uuid = Exec("wmic csproduct get UUID /value").Replace("UUID=", "").Trim();
                    return "vm:" + uuid;
                }
            }
            return "";
        }

        private static string OsType()
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) return "windows";
            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) return "macos";
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) return "linux";
            return Environment.OSVersion.Platform.ToString().ToLower();
        }
    }
}
